### PR TITLE
Remove VPC from CloudConnection before deleting the CC

### PR DESCRIPTION
**Pull request by hamzy**
The PowerNS folks are asking us to:
1) delete the VPC endpoint from the cloud connection
2) query the cloud connection to make sure the VPC endpoint has been removed from the list
3) delete the cloud connection when 2) is true
4) delete the VPC

### CODE REVIEW

**Comment by hamzy** [comment]
@mjturek @clnperez

File: pkg/destroy/powervs/cloudconnection.go [hunk 1]
```diff
@@ -1,24 +1,44 @@
...
+
+		// https://github.com/IBM-Cloud/power-go-client/blob/v1.0.88/power/models/cloud_connection_endpoint_v_p_c.go#L19-L26
+		EndpointVpc *models.CloudConnectionEndpointVPC
+
+		// https://github.com/IBM-Cloud/power-go-client/blob/v1.0.88/power/models/cloud_connection_v_p_c.go#L18-L26
+		Vpc *models.CloudConnectionVPC
+
+		foundOne       bool = false
+		foundVpc       bool = false
+		vpcStillExists bool = true
```
Comment (r4f4) [inline pkg/destroy/powervs/cloudconnection.go:38]: Nit: no need for the type; it's inferred from the rhs.

File: pkg/destroy/powervs/cloudconnection.go [hunk 2]
```diff
@@ -34,29 +54,73 @@ func (o *ClusterUninstaller) listCloudConnections() (cloudResources, error) {
...
-				id:       *jobReference.ID,
-			})
+				time.Sleep(15 * time.Second)
+			} else {
+				vpcStillExists = false
+			}
+		}
+
+		// Finally delete the CloudConnection!
+		jobReference, err = o.cloudConnectionClient.Delete(*cloudConnection.CloudConnectionID)
```
Comment (r4f4) [inline pkg/destroy/powervs/cloudconnection.go:110]: Nit:
```suggestion
		jobReference, err := o.cloudConnectionClient.Delete(*cloudConnection.CloudConnectionID)
```
Since this variable is only used here. No need to declare it a function level.

File: pkg/destroy/powervs/cloudconnection.go [hunk 3]
```diff
@@ -105,6 +169,8 @@ func (o *ClusterUninstaller) destroyCloudConnections() error {
 		if len(items) == 0 {
 			break
 		}
+
+		time.Sleep(15 * time.Second)
```
Comment (r4f4) [inline pkg/destroy/powervs/cloudconnection.go:173]: See previous comment.

File: pkg/destroy/powervs/dhcp.go [hunk 4]
```diff
@@ -30,16 +30,16 @@ func (o *ClusterUninstaller) listDHCPNetworks() (cloudResources, error) {
...
 		}
 		if dhcpServer.Network.Name == nil {
-			o.Logger.Debugf("listDHCPNetworks: DHCP has empty Network.Name: %s\n", *dhcpServer.ID)
+			o.Logger.Debugf("listDHCPNetworks: DHCP has empty Network.Name: %s", *dhcpServer.ID)
 			continue
 		}
 
 		if strings.Contains(*dhcpServer.Network.Name, o.InfraID) {
-			o.Logger.Debugf("listDHCPNetworks: FOUND: %s (%s)\n", *dhcpServer.Network.Name, *dhcpServer.ID)
+			o.Logger.Debugf("listDHCPNetworks: FOUND: %s (%s)", *dhcpServer.Network.Name, *dhcpServer.ID)
```
Comment (r4f4) [inline pkg/destroy/powervs/dhcp.go:42]: This is not related to the changes in the PR. It should at least be in a separate commit or on its own PR.

Comment (r4f4) [reply pkg/destroy/powervs/dhcp.go:42]: The same comment about structured logging is also valid here. Consider using `o.Logger.WithField`.

File: pkg/destroy/powervs/job.go [hunk 5]
```diff
@@ -131,6 +131,8 @@ func (o *ClusterUninstaller) destroyJobs() error {
 		if len(items) == 0 {
 			break
 		}
+
+		time.Sleep(15 * time.Second)
```
Comment (r4f4) [inline pkg/destroy/powervs/job.go:135]: See previous comment.

File: pkg/destroy/powervs/loadbalancer.go [hunk 6]
```diff
@@ -147,6 +150,8 @@ func (o *ClusterUninstaller) destroyLoadBalancers() error {
 		if len(items) == 0 {
 			break
 		}
+
+		time.Sleep(15 * time.Second)
```
Comment (r4f4) [inline pkg/destroy/powervs/loadbalancer.go:154]: See previous comment.

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+		Vpc *models.CloudConnectionVPC
+
+		// https://github.com/IBM-Cloud/power-go-client/blob/v1.0.88/power/models/cloud_connection_update.go#L20
+		cloudConnectionUpdate models.CloudConnectionUpdate
+
+		foundOne bool = false
+		foundVpc bool = false
+	)
+
+	ctx, _ = o.contextWithTimeout()
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:45]: See previous related comment.

File: pkg/destroy/powervs/cloudconnection.go [hunk 8]
```diff
@@ -1,24 +1,44 @@
...
+
+		// https://github.com/IBM-Cloud/power-go-client/blob/v1.0.88/power/models/cloud_connection_v_p_c.go#L18-L26
+		Vpc *models.CloudConnectionVPC
+
+		foundOne       bool = false
+		foundVpc       bool = false
+		vpcStillExists bool = true
+	)
+
+	ctx, _ = o.contextWithTimeout()
```
[the remark this answers is not in the document]
Comment (r4f4) [reply pkg/destroy/powervs/cloudconnection.go:41]: The docs for `context` say
```
Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires.
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	default:
+	}
+
+	cloudConnections, err = o.cloudConnectionClient.GetAll()
+	if err != nil {
+		log.Fatalf("Failed to list cloud connections: %v", err)
+	}
+
+	result := []cloudResource{}
+	for _, cloudConnection = range cloudConnections.CloudConnections {
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:62]: Nit:
```suggestion
	for _, cloudConnection := range cloudConnections.CloudConnections {
```
So there is no need to declare `var cloudConnection` at the beginning of the function.

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+		if err != nil {
+			log.Fatalf("Failed to get cloud connection %s: %v", cloudConnectionID, err)
+		}
+
+		endpointVpc = cloudConnection.Vpc
+
+		o.Logger.Debugf("listVPCInCloudConnections: endpointVpc = %+v", endpointVpc)
+
+		foundVpc = false
+		for _, Vpc = range endpointVpc.Vpcs {
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:91]:
```suggestion
		for _, Vpc := range endpointVpc.Vpcs {
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		o.Logger.Debugf("listVPCInCloudConnections: foundVpc = %v", foundVpc)
+		if !foundVpc {
+			continue
+		}
+
+		// https://github.com/IBM-Cloud/power-go-client/blob/v1.0.88/power/models/cloud_connection_v_p_c.go#L18
+		var vpcsUpdate []*models.CloudConnectionVPC
+
+		for _, Vpc = range endpointVpc.Vpcs {
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:106]:
```suggestion
		for _, Vpc := range endpointVpc.Vpcs {
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+			if !strings.Contains(Vpc.Name, o.InfraID) {
+				vpcsUpdate = append(vpcsUpdate, Vpc)
+			}
+		}
+
+		if len(vpcsUpdate) > 0 {
+			endpointUpdateVpc.Enabled = true
+		} else {
+			endpointUpdateVpc.Enabled = false
+		}
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:116]: Nit:
```suggestion
	    endpointUpdateVpc.Enabled = len(vpcsUpdate) > 0
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+		endpointUpdateVpc.Vpcs = vpcsUpdate
+
+		cloudConnectionUpdate.Vpc = &endpointUpdateVpc
+
+		var vpcsStrings []string
+
+		for _, Vpc = range vpcsUpdate {
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:124]:
```suggestion
		for _, Vpc := range vpcsUpdate {
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+			name:     *jobReference.ID,
+			status:   "",
+			typeName: jobTypeName,
+			id:       *jobReference.ID,
+		})
+	}
+
+	if !foundOne {
+		o.Logger.Debugf("listVPCInCloudConnections: NO matching cloud connections")
+		for _, cloudConnection = range cloudConnections.CloudConnections {
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:149]:
```suggestion
		for _, cloudConnection := range cloudConnections.CloudConnections {
```

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	)
+
+	found, err = o.listVPCInCloudConnections()
+	if err != nil {
+		return err
+	}
+
+	items = o.insertPendingItems(jobTypeName, found.list())
+
+	ctx, _ = o.contextWithTimeout()
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:174]: See previous related comment.

File: pkg/destroy/powervs/vpc-in-cloud-connection.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+				o.errorTracker.suppressWarning(item.key, err, o.Logger)
+			}
+		}
+
+		items = o.getPendingItems(jobTypeName)
+		if len(items) == 0 {
+			break
+		}
+
+		time.Sleep(15 * time.Second)
```
Comment (r4f4) [inline pkg/destroy/powervs/vpc-in-cloud-connection.go:202]: See previous related comment.

**Review by r4f4 — CHANGES_REQUESTED** [review]

File: pkg/destroy/powervs/cloudconnection.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hamzy) [reply pkg/destroy/powervs/cloudconnection.go:41]:
```
hamzy:hamzy-installer$ grep -r o.contextWithTimeout ./pkg/destroy/powervs/ | wc -l
38
```
Since this is all over the place, how about if I address this in a separate PR?

File: pkg/destroy/powervs/cloudconnection.go [hunk 17]
```diff
@@ -34,29 +54,73 @@ func (o *ClusterUninstaller) listCloudConnections() (cloudResources, error) {
...
+				o.Logger.Debugf("listCloudConnections: This CC still has VPCs attached, waiting...")
 
-			result = append(result, cloudResource{
-				key:      *jobReference.ID,
-				name:     *jobReference.ID,
-				status:   "",
-				typeName: jobTypeName,
-				id:       *jobReference.ID,
-			})
+				time.Sleep(15 * time.Second)
```
[the remark this answers is not in the document]
Comment (hamzy) [reply pkg/destroy/powervs/cloudconnection.go:103]: The sleeps are in a loop.  So without the sleep, we call PowerVS multiple times a second.

File: pkg/destroy/powervs/vpc-in-cloud-connection.go (hunk 9 above)
Comment (hamzy) [reply pkg/destroy/powervs/vpc-in-cloud-connection.go:62]: I felt a need.  That is why I went out of my way include not easy to find structure commenting.  This saves me a lot of work in tracking down where in the client is the particular structure definition.

File: pkg/destroy/powervs/dhcp.go (hunk 4 above)
Comment (hamzy) [reply pkg/destroy/powervs/dhcp.go:42]: Thanks for the pointer!  I will keep that in mind for a future PR!

Comment (hamzy) [reply pkg/destroy/powervs/dhcp.go:42]: The above comment was for the structured logging.  As for the \n, I was hoping to sneak them in and save a lot of work in having to do another PR which takes a crazy amount of time to get people to accept and push it.

File: pkg/destroy/powervs/cloudconnection.go (hunk 1 above)
Comment (hamzy) [reply pkg/destroy/powervs/cloudconnection.go:38]: Fixed

File: pkg/destroy/powervs/vpc-in-cloud-connection.go (hunk 12 above)
Comment (hamzy) [reply pkg/destroy/powervs/vpc-in-cloud-connection.go:116]: Fixed

**Comment by hamzy** [comment]
@r4f4 Did I successfully address your concerns?

File: pkg/destroy/powervs/cloudconnection.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (r4f4) [reply pkg/destroy/powervs/cloudconnection.go:103]: Why not use one of the functions from the https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait? Like https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollImmediateUntilWithContext or https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#ExponentialBackoffWithContext

File: pkg/destroy/powervs/dhcp.go (hunk 4 above)
Comment (r4f4) [reply pkg/destroy/powervs/dhcp.go:42]: Then at least sneak them in in their own commit explaining their purpose. If the commit for removing the VPC is ever reverted, this would be reverted too.

File: pkg/destroy/powervs/cloudconnection.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (r4f4) [reply pkg/destroy/powervs/cloudconnection.go:41]: Why not both? Since you're already adding these lines, add them with the `defer cancel()` and send a new PR fixing the other places.
